Add cross compiling functionality

dipinhora opened the pull request:
This pull request adds cross compiling functionality to ponyc. It does not address cross linking and the usual gcc cross compile chain is required for linking.

See: https://github.com/dipinhora/rpxp for an example of use of the cross compiling features in a Linux X86_64 environment targeting a Linux Arm 32 bit environment.

* Update all code generation related bits (codegen, expr, ast, pass,
  pkg) in libponyc to dynamically check for target machine features
  (OS, architecture, etc) based on LLVM Triple.
* Adjust Makefile for building to 32 bit targets on 64 bit platforms.
* Update test/libponyc/util to call `package_init` in order to
  properly intialize compiler for dynamic targets to fix some failing
  tests.
* Disable some tests in test/libponyc/program.cc because they fail
  when run as part of the full test suite but pass when run without
  other tests in the test suite
  (`./build/debug/libponyc.tests --gtest_also_run_disabled_tests --gtest_filter=ProgramTest.*`).
  These tests can be fixed by adding a `package_done` to `test/libponyc/util.cc` via the following diff:
```
diff --git a/test/libponyc/util.cc b/test/libponyc/util.cc
index 2fa46a2..5e7aa4b 100644
--- a/test/libponyc/util.cc
+++ b/test/libponyc/util.cc
@@ -345,6 +345,8 @@ void PassTest::build_package(const char* pass, const char* src,
 
     ASSERT_NE((void*)NULL, *out_package);
   }
+^M
+  package_done(&opt);^M
 }
```
  but this causes other tests to fail. I've confirmed that these same tests fail if the package_init/package_done are added to Pony master also so the failures are not related to the cross compiling changes.

The changes to `src/libponyc/codegen/genprim.c` are a bit of a hack in relation to the `num_conv_t conv[]` declaration/assignment due to the duplication.

andymcn commented:
Thanks for the huge effort!

When you say this doesn't address linking, do you mean that we're still just running the link command appropriate to the host machine? Presumably if the host linker can cope with the target platform then it will just work and otherwise it will produce some interesting error messages. I note that your specific case of interest is Intel 64 bit to ARM 32 bit both with Linux, so I assume this is easy enough to get working.

Unfortunately the tests have thrown up some issues with this PR.

1. On Linux, release builds are segfaulting when trying to build `packages/stdlib`.
1. On OS X, the unit test `BadPonyTest.ClassInOtherClassProvidesList` is now hitting an assert in the LLVM code:
"Assertion failed: (!ErrorHandler && "Error handler already registered!\n"), function install_fatal_error_handler, file /tmp/llvm20151214-56015-1irriwq/llvm-3.6.2.src/lib/Support/ErrorHandling.cpp, line 50."

dipinhora commented:
Hi Andy,

Thanks for the quick response.

When I say this doesn't address cross-linking, I mean that this patch doesn't change anything about how ponyc does linking and that if someone wanted to do a full cross compiling and cross linking setup they should be using ponyc to generate the object file and then manually do the linking with the appropriate cross linker. This is not to say that the default linker couldn't handle things properly or that ponyc (with additional logic) couldn't invoke the appropriate linker, just that at the moment, I didn't address it because I couldn't address it in the same general manner as cross compiling.

Yes, my specific scenario was to cross compile from Intel 64 bit to Arm 32 bit but I initially started out with Intel 32 bit to Arm 32 bit (since initially it was easier to ignore how to make a 64 bit ponyc generate 32 bit code). Realistically, the way things are set up in this pull request, ponyc can create the appropriate object file for any target that llvm supports (assuming you tell it where to find the libponyrt.a for the target platform) and the final cross linking (including installation of the appropriate cross linker and the libs for the target platform and the cross compile of libponyrt.a) are left to the person setting up the cross compile environment. https://github.com/dipinhora/rpxp/blob/master/Dockerfile#L6 has the details of everything (possibly more things than strictly required) needed for a cross compile environment to target armhf. https://github.com/dipinhora/rpxp/blob/master/Dockerfile#L41-L51 takes care of cross compiling libponyrt for armhf. https://github.com/dipinhora/rpxp/blob/master/Dockerfile#L41-L51 takes care of compiling ponyc for x86_64 (which can target any platform llvm supports). https://github.com/dipinhora/rpxp/blob/master/build.sh#L5 uses ponyc to create the object file targeting armhf for a pony program. https://github.com/dipinhora/rpxp/blob/master/build.sh#L7-L18 takes care of the final cross linking of the object file with libponyrt.a for armhf and the standard libs for armhf. It should be relatively straight forward to set up a similar cross compile environment for any other target with the appropriate cross compiler and cross linkers.

In regards to the failing builds in Travis CI:

1. I looked into the release builds segfaulting on Linux x86_64 (I had done all my work on the debug builds so didn't even realize this was an issue until the Travis build failed) and I am encountering something peculiar. The normal release build does a `-O3` and it segfaults but if I replace it with `-O0 -<individual optimization flags>` it works. `<individual optimization flags>` is the output of `gcc -Q --help=optimizers -O3 > /tmp/O3 && gcc -Q --help=optimizers -O0 > /tmp/O0 && diff /tmp/O0 /tmp/O3| grep '^>'| awk '{print $2 " \\"}'`. This is highly confusing to me since my understanding is that both of those should be equivalent. I had hoped instead to be able to disable optimizations in order to narrow down which specific (or combination of) optimization(s) is causing the segfault. Any ideas?
2. I haven't looked into the OS X build issues yet but will try to in the next few days.
3. Regarding the clang failures, I'm able to reproduce the issue on Linux x86_64 but unlike gcc I'm not able to debug further as I don't know how to manually reproduce the additional optimizations that `-O3` applies that aren't applied normally so I can try to narrow down which specific optimization is causing the segfault.
4. As an aside regarding the ponyc tests, the current code in master (and this pull request) for `PassTest::build_package` in `util.cc` doesn't do a `package_init` and `package_done` (not in current pull request) but once these are added, some tests start failing and others start behaving oddly (test passes but output indicates error). This is an issue because for the cross compiling support both `package_init` and `package_done` are required in order to properly initialize/tear down the llvm compilation environment. This is true even when doing a native build because all code generation is now based on the llvm target triple which has to be determined at runtime. Unfortunately, as of right now, I do not know enough C/C++ (I haven't done anything "real" in over 15 years until this) nor enough about Pony to understand/fix these but hopefully you can.

Let me know if you have any additional questions. I'll post a followup comment once I have looked more into the OS X build issue.

andymcn commented:
Don't worry about not thinking to check the release build, we used to do that a lot. There are many commits in this repo with the comment "Fix release build" or similar.

I'd actually approach debugging the release build segfault completely differently. Rather than trying to work out which option causes the problem I'd find out where in the code the problem occurs.

You should be able to reproduce the crash just by running (with a release build of ponyc):
```
ponyc packages/stdlib
```

Run this inside gdb (or lldb if you prefer) and you'll get an instant back trace that will hopefully show you exactly where things are dying and what the relevant arguments, local variables, etc are.

Since the clang and gcc tests are both failing with segfaults in the code generation pass I'd guess they're the same problem.

Typically when things work fine in debug but segfault in release it's due to memory not being 0'ed in release and hence a random piece of data being treated as a pointer and dereferenced.

For the unit test `package_init` issues, sorry that's almost certainly my fault. Just disable (but not delete) the unit tests you need to due to this and I'll fix things once you're merged.

dipinhora commented:
Thanks for the pointers. I had initially started with gdb but didn't get very far and had given up to look for other approaches (my lack of experience with C/C++ definitely didn't help). After your last comment, I went back to gdb/lldb and print statements and identified something weird but couldn't find the root cause after many hours of debugging/tracing.

Luckily, I was able to get @sylvanc to take a quick look in person and 15 minutes later he had identified my error. See the commit with message `Fix lexical scoping issue to resolve release build issues.` for the details.

I also re-enabled all tests and fixed the issues so now all tests pass with these changes.

Let me know if you have any questions or concerns or comments regarding how I implemented things.
